junit-platform-launcher/src/main/java9/org/junit/platform/launcher/listeners/FlightRecordingListener.java

```diff
@@ -0,0 +1,180 @@
+import org.junit.platform.engine.reporting.ReportEntry;
+import org.junit.platform.launcher.TestExecutionListener;
+import org.junit.platform.launcher.TestIdentifier;
+import org.junit.platform.launcher.TestPlan;
+
+/**
+ * A {@link TestExecutionListener} that generates Java Flight Recorder
+ * events.
+ *
+ * @see <a href="https://openjdk.java.net/jeps/328">JEP 328: Flight Recorder</a>
+ * @since 1.7
+ */
+@API(status = EXPERIMENTAL, since = "1.7")
+public class FlightRecordingListener implements TestExecutionListener {
+
+	private final AtomicReference<TestPlanExecutionEvent> testPlanExecutionEvent = new AtomicReference<>();
+	private final Map<String, TestExecutionEvent> testExecutionEventMap = new ConcurrentHashMap<>();
+
+	@Override
+	public void testPlanExecutionStarted(TestPlan plan) {
+		TestPlanExecutionEvent event = new TestPlanExecutionEvent();
+		event.containsTests = plan.containsTests();
+		event.engineNames = plan.getRoots().stream().map(TestIdentifier::getDisplayName).collect(
+			Collectors.joining(", "));
+		testPlanExecutionEvent.set(event);
+		event.begin();
+	}
+
+	@Override
+	public void testPlanExecutionFinished(TestPlan plan) {
+		TestPlanExecutionEvent event = testPlanExecutionEvent.get();
+		event.commit();
+	}
+
+	@Override
+	public void executionSkipped(TestIdentifier test, String reason) {
+		SkippedTestEvent event = new SkippedTestEvent();
+		event.initialize(test);
+		event.reason = reason;
+		event.commit();
+	}
+
+	@Override
+	public void executionStarted(TestIdentifier test) {
+		TestExecutionEvent event = new TestExecutionEvent();
+		testExecutionEventMap.put(test.getUniqueId(), event);
+		event.initialize(test);
+		event.begin();
+	}
+
+	@Override
+	public void executionFinished(TestIdentifier test, TestExecutionResult result) {
+		if (test.isContainer() && result.getStatus().equals(TestExecutionResult.Status.SUCCESSFUL)) {
+			return;
+		}
+		Optional<Throwable> throwable = result.getThrowable();
+		TestExecutionEvent event = testExecutionEventMap.get(test.getUniqueId()); // TODO Remove?
+		event.end();
+		event.result = result.getStatus().toString();
+		event.exceptionClass = throwable.map(Throwable::getClass).orElse(null);
+		event.exceptionMessage = throwable.map(Throwable::getMessage).orElse(null);
+		event.commit();
+	}
+
+	@Override
+	public void reportingEntryPublished(TestIdentifier test, ReportEntry reportEntry) {
+		for (Map.Entry<String, String> entry : reportEntry.getKeyValuePairs().entrySet()) {
+			ReportEntryEvent event = new ReportEntryEvent();
+			event.uniqueId = test.getUniqueId();
+			event.key = entry.getKey();
+			event.value = entry.getValue();
+			event.commit();
+		}
+	}
+
+	@Category("JUnit")
+	@Label("Test Plan")
+	@Name("org.junit.TestPlan")
+	@StackTrace(false)
+	static class TestPlanExecutionEvent extends Event {
+		@Label("Contains Tests")
+		boolean containsTests;
+		@Label("Engine Names")
+		String engineNames;
+	}
+
+	@MetadataDefinition
+	@Relational
```

Comment:
Set a name for the annotation, i.e @Name("org.junit.UniqueId")

Reply:
Sure.

Removed the name annotation from the metadata definition because it isn't inherited:

```java
@Target({ ElementType.TYPE, ElementType.FIELD, ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
@MetadataDefinition
public @interface Name { ...
```